src/csharp/Microsoft.Spark/Pyrolite/BufferUnpickler.cs

```diff
@@ -0,0 +1,859 @@
+// Licensed to the .NET Foundation under one or more agreements.
```

Comment:
Should we contribute this to https://github.com/irmen/Pyrolite? Note that we copied Pyrolite code because the dll was not being signed with a strong name as of last week. Now that it is being signed, we plan to remove Pyrolite code from our repo.

Reply:
It depends on how fast we want to go.

The BufferUnpickler introduced in this PR offers the same features as the original Unipicker. 

The thing is that we don't need all of the features it offers and we could make it tailored to our needs and even faster.

An example: the Pickler uses internally a stack of deserialized objects. If we want to deserialize a hundred double pairs (two doubles), it's going to allocate a hundred of `object[2]` arrays with two doubles in each array and store them on the mentioned stack. It's ok for general purpose.

But in our case, the `SqlCommandExecutor` always iterates over the deserialized `IEnumerable<object[]>` in a loop and it uses only one element at a time. So instead of allocating `new object[2]` for every pair of doubles, we could have one and make the pickler return a custom `IEnumerable<object[]>` which would be reusing the cached array

```cs
object[] inputRows = PythonSerDe.GetUnpickledObjects(new ReadOnlySpan<byte>(buffer, 0, messageLength));
for (int i = 0; i < inputRows.Length; ++i)
    outputRows.Add(commandRunner.Run(0, inputRows[i]));
```

Could become:

```cs
IEnumerable<object[]> inputRows = PythonSerDe.GetUnpickledObjects(new ReadOnlySpan<byte>(buffer, 0, messageLength));

foreach(object[] input row in inputRows)
    outputRows.Add(commandRunner.Run(0, inputRow)); // input row would be always the same instance of array (depending on size), but pointing to different values
```

Reply:
Most probably I should contribute it to the Pyrolite now and in the future if we decided to go the extreme path, imlement a tailored solution.